### ISSUE TITLE
Empty space filled with `q` characters under FreeBSD vtty

### ISSUE THREAD

**Issue by mqudsi** [neovim/neovim#8644]
- `nvim --version`: 0.3.1 (master)
- Vim (version: 8.0.1638) behaves differently? yes
- Operating system/version: FreeBSD 11.1
- Terminal name/version: physical console
- `$TERM`: xterm

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
i
```

### Actual behaviour
The position under the cursor after `i` is actually empty, but a `q` is displayed (inverted under the grey cursor).

Same behavior is observed when pressing `:`,  the console shows `:q▋` 

CTRL+L clears the screen of these extraneous `q`s, but they continuously appear while navigating/editing.

### Expected behaviour
That they don't show.

**Comment by justinmk** [comment]
xterm can't be correct since those consoles apparently do not support cursor shaping (unlike the Linux VT console). Does bsd set TERM=xterm for those, or is it coming from your config?

**Comment by mqudsi** [comment]
It's the default value.

**Comment by mqudsi** [comment]
btw, what do you mean by cursor shaping? FreeBSD supports the `CONS_SETCURSORSHAPE` ioctl to set the cursor to normal, blinking, or hidden. But I don't think it supports changing it to a beam/pipe, if that's what you mean?

**Comment by justinmk** [comment]
> But I don't think it supports changing it to a beam/pipe, if that's what you mean?


Yes, that. So we need a way to detect the bsd console, since it claims to implement xterm but actually doesn't, and furthermore barfs unrecognized termcodes onto the screen instead of ignoring them.

**Comment by justinmk** [comment]
@mqudsi FWIW a workaround is described in https://github.com/neovim/neovim/wiki/FAQ#nvim-shows-weird-symbols-2-q-when-changing-modes .

Nvim enables cursor-shaping by default, except for some known uncooperative terminals. In this case it seems we don't have an easy way to detect the terminal. Ideally we would use terminal request-response to detect if the terminal supports cursor-shaping. https://github.com/neovim/neovim/issues/6279

**Comment by mission712** [comment]
I also had Konsole do this for me, not sure if that is fixed.

**Comment by mission712** [comment]
Hmm... Strange. It doesn't happen with nvim v0.3.1-85-g7ae7da8fb on Konsole, but it happens with ssh to my VPN which has v0.2.2.
Locally, $TERM is xterm-256color
via ssh, $TERM is also xterm-256color
maybe ssh is messing with is somehow? Or was the fix included after 0.2.2? I will try building on my server the latest version and see if it also pukes `q`s.

**Comment by justinmk** [comment]
Konsole is a different topic, already covered in other issues.

**Comment by justinmk** [comment]
@mission712 No. It's already covered in numerous other issues. Please search existing issues: https://github.com/neovim/neovim/issues?utf8=%E2%9C%93&q=is%3Aissue++konsole+ssh+

**Comment by mqudsi** [comment]
Hmm, I'm not seeing this behavior with FreeBSD 12-CURRENT (the prerelease builds) under VMware with neovim 0.3.0, but confusingly at the same time, it too declares itself as `xterm` and cursor shaping still doesn't happen.

**Comment by justinmk** [comment]
@mqudsi Maybe FreeBSD 12 changed its logic to ignore unrecognized sequences. That's becoming more common in terminals (and helps avoid all of the unreliable feature-detection voodoo that every terminal application is expected to reinvent).

**Comment by justinmk** [comment]
@mqudsi On your BSD system is TERM typically set to `xterm-256color` for GUI terminal apps? If so, then we can reasonably guess that `TERM=xterm` is a degraded VT. I'm going to try using that as a clue.

**Comment by mqudsi** [comment]
Yes, it is indeed. Smart approach.

**Comment by justinmk** [comment]
@mqudsi quick question: does tmux "hide" the problems reported here? If not, then I need to account for that in the logic.

**Comment by mqudsi** [comment]
@justinmk Had an opportunity to test it, and for the record stock FreeBSD installation has `TERM` set to `xterm` at the vty and `TERM` set to `screen` when running under a clean installation of `tmux` at the vty, so you should be OK with the patch.
